src/circuit/gadget/ecc/chip/mul/complete.rs

```diff
@@ -0,0 +1,186 @@
+    poly::Rotation,
+};
+
+use pasta_curves::{arithmetic::FieldExt, pallas};
+
+pub struct Config {
+    // Selector used to constrain the cells used in complete addition.
+    q_mul_decompose_var: Selector,
+    // Advice column used to decompose scalar in complete addition.
+    pub z_complete: Column<Advice>,
+    // Permutation
+    perm: Permutation,
+    // Configuration used in complete addition
+    add_config: add::Config,
+}
+
+impl From<&EccConfig> for Config {
+    fn from(ecc_config: &EccConfig) -> Self {
+        let config = Self {
+            q_mul_decompose_var: ecc_config.q_mul_decompose_var,
+            z_complete: ecc_config.advices[9],
+            perm: ecc_config.perm.clone(),
+            add_config: ecc_config.into(),
+        };
+
+        let add_config_advices = config.add_config.advice_columns();
+        assert!(
+            !add_config_advices.contains(&config.z_complete),
+            "z_complete cannot overlap with complete addition columns."
+        );
+
+        config
+    }
+}
+
+impl Config {
+    /// Gate used to check scalar decomposition is correct.
+    /// This is used to check the bits used in complete addition, since the incomplete
+    /// addition gate (controlled by `q_mul`) already checks scalar decomposition for
+    /// the other bits.
+    pub(super) fn create_gate(&self, meta: &mut ConstraintSystem<pallas::Base>) {
+        meta.create_gate(
+            "Decompose scalar for complete bits of variable-base mul",
+            |meta| {
+                let q_mul_decompose_var = meta.query_selector(self.q_mul_decompose_var);
+                // z_{i + 1}
+                let z_prev = meta.query_advice(self.z_complete, Rotation::prev());
+                // z_i
+                let z_next = meta.query_advice(self.z_complete, Rotation::next());
+
+                // k_{i} = z_{i} - 2⋅z_{i+1}
+                let k = z_next - Expression::Constant(pallas::Base::from_u64(2)) * z_prev;
+                let k_minus_one = k.clone() - Expression::Constant(pallas::Base::one());
+                // (k_i) ⋅ (k_i - 1) = 0
+                let bool_check = k.clone() * k_minus_one.clone();
+
+                // base_y
+                let base_y = meta.query_advice(self.z_complete, Rotation::cur());
+                // y_p
+                let y_p = meta.query_advice(self.add_config.y_p, Rotation::prev());
+
+                // k_i = 0 => y_p = -base_y
+                // k_i = 1 => y_p = base_y
+                let y_switch = k_minus_one * (base_y.clone() + y_p.clone()) + k * (base_y - y_p);
+
+                std::array::IntoIter::new([("bool_check", bool_check), ("y_switch", y_switch)])
+                    .map(move |(name, poly)| (name, q_mul_decompose_var.clone() * poly))
+            },
+        );
+    }
+
+    #[allow(clippy::type_complexity)]
+    #[allow(non_snake_case)]
+    #[allow(clippy::too_many_arguments)]
```

Comment:
Opinion: the default threshold of at most 7 arguments to avoid this lint is silly. We should raise that globally.